apps/arweave/src/ar_shutdown_manager.erl

```diff
@@ -0,0 +1,598 @@
+-export([handle_call/3, handle_cast/2, handle_info/2]).
+-export([
+	connections/0,
+	connections/1,
+	list_connections/0,
+	list_connections/1,
+	socket_info/1,
+	socket_info/2,
+	start_killer/1,
+	terminate_connections/0
+]).
+-include_lib("eunit/include/eunit.hrl").
+-include("ar_config.hrl").
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+start_link() ->
+	start_link(#{}).
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+start_link(Args) ->
+	gen_server:start_link({local, ?MODULE}, ?MODULE, Args, []).
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+init(_Args) ->
+	erlang:process_flag(trap_exit, true),
+	StartedAt = erlang:system_time(),
+	?LOG_INFO([{start, ?MODULE}, {pid, self()}, {started_at, StartedAt}]),
+	{ok, #{ started_at => StartedAt }}.
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+handle_call(uptime, _From, State = #{ started_at := StartedAt }) ->
+	Now = erlang:system_time(),
+	{reply, Now-StartedAt, State};
+handle_call(Msg, From, State) ->
+	?LOG_WARNING([{message, Msg}, {from, From}]),
+	{noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+handle_cast(Msg, State) ->
+	?LOG_WARNING([{message, Msg}]),
+	{noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @hidden
+%%--------------------------------------------------------------------
+handle_info(Msg, State) ->
+	?LOG_WARNING([{message, Msg}]),
+	{noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @hidden
+%% @doc this function is called when ar_sup is being stopped. If it
+%% was correctly configured, this should be the last function to be
+%% executed in the supervision tree.
+%% @end
+%%--------------------------------------------------------------------
+terminate(_Reason = shutdown, _State = #{ started_at := StartedAt }) ->
+	Now = erlang:system_time(),
+	terminate_connections(),
+	?LOG_INFO([
+		{stop, ?MODULE},
+		{started_at, StartedAt},
+		{stopped_at, Now},
+		{uptime, Now-StartedAt}
+	]).
+
+%%--------------------------------------------------------------------
+%% @hidden
+%% @doc terminate all active http connections from ranch/cowboy. This
+%% @end
+%%--------------------------------------------------------------------
+terminate_connections() ->
+	% this process should not exit, it will be linked to another
+	% process in charge to kill all connections.
+	erlang:process_flag(trap_exit, true),
```

Comment:
why do we need to call this twice (both here and during init/0)? (asking to understand, not to question)

Reply:
the `erlang:process_flag(trap_exit, true),` is still called in `terminate_connections` and `init` - is this intentional?